api/src/main/java/org/openmrs/module/appframework/context/ProgramConfiguration.java

```diff
@@ -0,0 +1,415 @@
+	 *  the {@code workflow} is also associated with the {@code program}.
+	 * 
+	 * @return {@code true} if a valid program tree was found
+	 */
+	public boolean hasValidProgramTree() {
+		Program program = getProgram();
+		ProgramWorkflow workflow = getWorkflow();
+		ProgramWorkflowState state = getState();
+		// Only program was specified
+		if (program != null && workflow == null && state == null) {
+			return true;			
+		}
+		// Only workflow was specified
+		if (program == null && workflow != null && state == null) {
+			return true;			
+		}
+		// Only state was specified
+		if (program == null && workflow == null && state != null) {
+			return true;			
+		}
+		// For cases where only workflow and state were specified, be sure the state belongs to the specified workflow
+		if (program == null && workflow != null && state != null) {
+			return workflow.getStates().contains(state);
+		}
+		// For cases where only program and state were specified, be sure the state belongs to a workflow associated with program
+		if (program != null && workflow == null && state != null) {
+			boolean stateInProgramTree = false;
+			for (ProgramWorkflow candidate : program.getAllWorkflows()) {
+				if (candidate.getStates().contains(state)) {
+					stateInProgramTree = true;
+				}
+			}
+			return stateInProgramTree;
+		}
+		// For cases where a workflow and program were specified, be sure the workflow belongs to the specified program
+		if (program != null && workflow != null) {
+			boolean programHasWorkflow = program.getAllWorkflows().contains(workflow);
+			if (state != null) {
+				// For cases where a workflow, program and state were specified
+				return programHasWorkflow && workflow.getStates().contains(state);
+			}
+			return programHasWorkflow;
+		}
+		return false;
+	}
+	
+	protected ResolvedConfiguration getResolvedConfig() {
+		if (resolvedConfig != null) {
+			return resolvedConfig;
+		}
+		ResolvedConfiguration ret = new ResolvedConfiguration();	    
+		if (StringUtils.isNotBlank(programRef)) {
+		    List<Program> programs = getAllPossiblePrograms();
+			if (programs.size() == 1) {
+				ret.setProgram(programs.get(0));
+			} else if (programs.size() > 1) {
+				ret.setProgram(programWithBestWorflowAndStateCombination(programs));
+			}
+		}			    	    
+		if (StringUtils.isNotBlank(workflowRef)) {
+		    List<ProgramWorkflow> workflows = getAllPossibleWorkflows();
+		    if (workflows.size() == 1) {
+				ret.setWorkflow(workflows.get(0));
+			} else if (workflows.size() > 1) {
+				ret.setWorkflow(workflowWithBestProgramAndStateCombination(workflows, ret.getProgram()));
+			} 
+		}		
+		if (StringUtils.isNotBlank(stateRef)) {
+		    List<ProgramWorkflowState> states = getAllPossibleStates();
+		    if (states.size() == 1) {
+				ret.setState(states.get(0));
+			} else if (states.size() > 1) {
+				ret.setState(stateWithBestProgramAndWorkflowCombination(states, ret.getProgram(), ret.getWorkflow()));
+			}
+		}	    
```

Comment:
At this point you have three shortlists:
- `programs`
- `workflows` (that are working with `programs`)
- `states` (that are working with `workflows`)

And this is where you need a specialized routine that will take the three shortlists and reduce them to a single final triplet program/workflow/state _or_ throw an `IllegalArgumentException` if  that is not possible from the provided references.

Probably this existing `stateWithBestProgramAndWorkflowCombination` to be renamed `findProgramWorkflowStateTriplet`.